src/main/java/bio/terra/service/dataset/AzureStorageResource.java

```diff
@@ -0,0 +1,88 @@
+package bio.terra.service.dataset;
+
+import bio.terra.app.model.AzureCloudResource;
+import bio.terra.app.model.AzureRegion;
+import bio.terra.app.model.CloudRegion;
+import bio.terra.app.model.CloudResource;
+import bio.terra.model.CloudPlatform;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+
+import java.util.Objects;
+import java.util.UUID;
+
+@JsonTypeName("azure")
+public class AzureStorageResource implements StorageResource {
+
+    private UUID datasetId;
+    private CloudPlatform cloudPlatform;
+    private AzureCloudResource cloudResource;
+    private AzureRegion region;
+
+    AzureStorageResource() {
+        this.cloudPlatform = CloudPlatform.AZURE;
+    }
+
+    @Override
+    public UUID getDatasetId() {
+        return datasetId;
+    }
+
+    @Override
+    public AzureStorageResource datasetId(UUID datasetId) {
+        this.datasetId = datasetId;
+        return this;
+    }
```

Comment:
Do we need fluent style setters for this class? I'd really like to migrate away from mutable state in our objects in general, and switch to using builder-style construction or multi-arg constructors.

Reply:
That's a broader question.  I agree but if we do that, I'd like to do a wholesale switchover

Reply:
I disagree about the wholesale requirement, especially in a case like this where the change is specific to an individual class. In this case it's a new class in this PR, so it's an even better time to try doing things a different way IMO.

Reply:
I think if we start using builders, I'd like to do it at a larger scale and also figure out what details about builder we want (e.g. separate class, same class, etc.)

Reply:
In this case I don't think a builder is needed, as you always have all the arguments when you construct one. I think most cases where we use fluent-style setters could be written as constructor calls. There are probably a few cases where there's enough variation in construction that a builder would make sense, but in most cases I would bet we don't need to support all the combinations of setting one field or another.

Reply:
I was thinking, another reason why the field should be `final` is because this class has hashCode(). If you make a collection with this class and then modify it, its hashcode and equality will change out from under it and the collection state could become invalid. In general we only want to make hashed (or ordered) collections over immutable data. Unless the collection is aware of the mutation and can reorganize itself it its items state changfes.